common/nation.cpp

```diff
@@ -455,166 +455,80 @@ struct nation_type *nation_by_number(const Nation_type_id nation)
-  it->p = nations;
-  it->end = nations + nation_count();
-  return ITERATOR(it);
+  return pnation - &nations[0];
 }
 
 /**
    Allocate resources associated with the given nation.
+   Constructor for nation_type
  */
-static void nation_init(struct nation_type *pnation)
+nation_type::nation_type()
 {
-  memset(pnation, 0, sizeof(*pnation));
-
-  pnation->item_number = pnation - nations;
-  pnation->translation_domain = nullptr;
-  pnation->leaders = nation_leader_list_new_full(nation_leader_destroy);
-  pnation->sets = nation_set_list_new();
-  pnation->groups = nation_group_list_new();
+  item_number = this - &nations[0];
+  translation_domain = nullptr;
+  leaders = nation_leader_list_new_full(nation_leader_destroy);
+  sets = nation_set_list_new();
+  groups = nation_group_list_new();
 
   if (is_server()) {
-    pnation->server.default_cities =
-        nation_city_list_new_full(nation_city_destroy);
-    pnation->server.civilwar_nations = nation_list_new();
-    pnation->server.parent_nations = nation_list_new();
-    pnation->server.conflicts_with = nation_list_new();
+    server.default_cities = nation_city_list_new_full(nation_city_destroy);
+    server.civilwar_nations = nation_list_new();
+    server.parent_nations = nation_list_new();
+    server.conflicts_with = nation_list_new();
     // server.rgb starts out nullptr
-    pnation->server.traits = new trait_limits[TRAIT_COUNT];
+    server.traits = new trait_limits[TRAIT_COUNT];
   }
 }
 
 /**
    De-allocate resources associated with the given nation.
+   Destructor for nation_type
  */
-static void nation_free(struct nation_type *pnation)
+nation_type::~nation_type()
 {
-  delete[] pnation->legend;
-  delete[] pnation->translation_domain;
-  nation_leader_list_destroy(pnation->leaders);
-  nation_set_list_destroy(pnation->sets);
-  nation_group_list_destroy(pnation->groups);
+  delete[] legend;
+  delete[] translation_domain;
+  nation_leader_list_destroy(leaders);
+  nation_set_list_destroy(sets);
+  nation_group_list_destroy(groups);
 
   if (is_server()) {
-    nation_city_list_destroy(pnation->server.default_cities);
-    nation_list_destroy(pnation->server.civilwar_nations);
-    nation_list_destroy(pnation->server.parent_nations);
-    nation_list_destroy(pnation->server.conflicts_with);
-    rgbcolor_destroy(pnation->server.rgb);
-    delete[] pnation->server.traits;
+    nation_city_list_destroy(server.default_cities);
+    nation_list_destroy(server.civilwar_nations);
+    nation_list_destroy(server.parent_nations);
+    nation_list_destroy(server.conflicts_with);
+    rgbcolor_destroy(server.rgb);
+    delete[] server.traits;
   }
-
-  memset(pnation, 0, sizeof(*pnation));
 }
 
 /**
    Allocate space for the given number of nations.
  */
 void nations_alloc(int num)
 {
-  int i;
-
-  nations = new nation_type[num];
   game.control.nation_count = num;
-
-  for (i = 0; i < num; i++) {
-    nation_init(nations + i);
-  }
+  nations.resize(game.control.nation_count);
```

Comment:
`game.control.nation_count` => `nations.size()`? (possibly wrapped in a function)

Reply:
No, the vector is size 0 until it is resized here. I'd keep game.control.nation_count this way, atleast for now.